File: Cpp/Merge k Sorted Lists.cpp
#include <vector>
#include <queue>

/**
 * Definition of ListNode
 */
class ListNode {
public:
    int val;
    ListNode *next;
    ListNode(int val) {
        this->val = val;
        this->next = NULL;
    }
};

class Solution {
public:
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    struct cmp{
        bool operator ()(const ListNode *a, const ListNode *b){
            return a->val > b->val;
        }
    };
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        // write your code here
        priority_queue<ListNode*, vector<ListNode*>, cmp> que;
        ListNode *pre = new ListNode(-1), *cnt = pre;
        for (int i = 0; i < lists.size(); i ++) {
            if (lists[i] != NULL) {
                que.push(lists[i]);
            }
        }
        while (!que.empty()) {
            cnt->next = que.top();
            cnt = cnt->next;
            if(que.top()->next != NULL) que.push(que.top()->next);
            que.pop();
        }
        cnt->next = NULL;
        return pre->next;
    }
};

File: Java/Merge-k-Sorted-Lists.java
public class Solution {
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        // write your code here
        ListNode dummy = new ListNode(-1);
        ListNode cur = dummy;
        while (l1 != null && l2 != null) {
            if (l1.val <= l2.val) {
                cur.next = l1;
                cur = cur.next;
                l1 = l1.next;
            } else {
                cur.next = l2;
                cur = cur.next;
                l2 = l2.next;
            }
        }
        if (l1 == null) {
            cur.next = l2;
        } else {
            cur.next = l1;
        }
        return dummy.next;
    }
    public ListNode mergeHelper(List<ListNode> lists, int start, int end) {
        if (start == end) return lists.get(start);
        int mid = start + (end - start) / 2;
        ListNode left = mergeHelper(lists, start, mid);
        ListNode right = mergeHelper(lists, mid + 1, end);
        return mergeTwoLists(left, right);
    }
    public ListNode mergeKLists(List<ListNode> lists) {
        // write your code here
        if (lists.size() == 0) return null;
        return mergeHelper(lists, 0, lists.size() - 1);
    }
}